lib/discourse_auth_cookie.rb

```diff
@@ -0,0 +1,137 @@
+# frozen_string_literal: true
+
+class DiscourseAuthCookie
+  class InvalidCookie < StandardError; end
+
+  class Encryptor
+    def encrypt_and_sign(plain_cookie)
+      encryptor.encrypt_and_sign(plain_cookie)
+    end
+
+    def decrypt_and_verify(cipher_cookie)
+      encryptor.decrypt_and_verify(cipher_cookie)
+    rescue ActiveSupport::MessageVerifier::InvalidSignature
+      raise InvalidCookie.new
+    end
+
+    private
+
+    def encryptor
+      @encryptor ||= begin
+        key = Rails.application.key_generator.generate_key("discourse-auth-cookie", 32)
```

Comment:
Doesn't the salt have to be unique per application and be kept secret? Otherwise, the key generated will be same between all applications and anyone has the key to decrypt the auth cookie.

Reply:
`Rails.application.key_generator` derives a key from `Rails.application.config.secret_key_base` which is unique per install so the derived key is also unique per install. I don't think having a unique/secret salt adds any extra security, does it? The base key is already unique and randomly generated and in case of a breach both the base key and salt would be leaked and therefore it'd be quite straightforward to derive the final key?

Also Rails lets us set custom salt for the session key via `Rails.application.config.action_dispatch` but we don't change the default salt.

Will go the rest of your comments today, thanks for the thorough review!

Reply:
> `Rails.application.key_generator` derives a key from `Rails.application.config.secret_key_base` which is unique per install so the derived key is also unique per install.

Ooo if it relies on `secret_key_base` then we're good here. I was only reading through the `.generate_key` method and it wasn't obvious that the `secret_key_base` is used anywhere.

Reply:
I think we should include the database name in the salt. Otherwise, a cookie for one site in a multisite cluster could be used to bypass the rate limiter for all other sites in the cluster.

Reply:
Hmm do we not use a different secret key base for each site per multi-site already? I weirdly have some impression that this is already the case because the encryption of sessions rely on the secret key base.

Reply:
I can't find anything which would change the secret per-site. Previously it didn't matter much because we always looked-up the `_t` token in the site's own database. Now that we're relying on the cryptography, it becomes much more important.

You're absolutely right that we should fix up the `session` cookie as well, so that it has a uniquely salted secret per-site :+1: (but maybe that should be a separate PR)

Reply:
Just to confirm, the salt doesn't have to be kept secret, correct? So each site in a multisite clusters gets its own unique salt, but keeping the salt secret doesn't have any advantages?

Reply:
Yes I think that's correct - the salt itself doesn't need to be secret.